programs/mpl-core/src/plugins/lifecycle.rs

```diff
@@ -654,6 +807,124 @@ pub(crate) trait PluginValidation {
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the remove plugin lifecycle action.
+    /// This gets called on all existing plugins when the target plugin is removed.
+    fn side_effects_remove_plugin(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the add plugin lifecycle action.
+    /// This gets called on all existing plugins when a new external plugin is added.
+    fn side_effects_add_external_plugin_adapter(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the remove plugin lifecycle action.
+    /// This gets called on all existing plugins when a new external plugin is removed.
+    fn side_effects_remove_external_plugin_adapter(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the approve plugin authority lifecycle action.
+    fn side_effects_approve_plugin_authority(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the revoke plugin authority lifecycle action.
+    fn side_effects_revoke_plugin_authority(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the create lifecycle action.
+    /// This ONLY gets called to side_effects the self plugin
+    fn side_effects_create(&self, _ctx: &PluginSideEffectContext) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the update lifecycle action.
+    /// This gets called on all existing plugins when an asset or collection is updated.
+    fn side_effects_update(&self, _ctx: &PluginSideEffectContext) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the update_plugin lifecycle action.
+    /// This gets called on all existing plugins when a plugin is updated.
+    fn side_effects_update_plugin(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the burn lifecycle action.
+    /// This gets called on all existing plugins when an asset is burned.
+    fn side_effects_burn(&self, _ctx: &PluginSideEffectContext) -> Result<Plugin, ProgramError> {
+        unreachable!()
+    }
+
+    /// Process side effects for the transfer lifecycle action.
+    /// This gets called on all existing plugins when an asset is transferred.
+    fn side_effects_transfer(
+        &self,
+        _ctx: &PluginSideEffectContext,
+    ) -> Result<Plugin, ProgramError> {
```

Comment:
wdyt about the function prototype returning `Result<Self, ProgramError>`.  Are there cases where we should operate on another plugin, or is it more like it should just operate on itself?